IPython/core/history.py

```diff
@@ -880,10 +888,10 @@ def __init__(self, history_manager):
         super(HistorySavingThread, self).__init__(name="IPythonHistorySavingThread")
         self.history_manager = history_manager
         self.enabled = history_manager.enabled
-        atexit.register(self.stop)
 
     @only_when_enabled
     def run(self):
+        atexit.register(self.stop)
```

Comment:
Hum, interesting. I haven't looked at this code in while, and not too familiar with threading.

Would it also be more correct to `atexit.unregister(self.stop)` in `def stop(self)` itself ?

Reply:
Seems reasonable, done in 047ab6e